web/packages/teleport/src/Integrations/Enroll/MachineId/GitHub/useGitHubFlow.tsx

```diff
@@ -0,0 +1,258 @@
+                  return {
+                    repository: `${owner}/${repository}`,
+                    repository_owner: owner,
+                    actor: r.actor,
+                    environment: r.environment,
+                    ref: r.ref,
+                    ref_type: r.refType.value || null,
+                    workflow: r.workflowName,
+                  };
+                }),
+              },
+            })
+            .then(token => {
+              setTokenName(token.id);
+              return botService.createBot(botConfig);
+            });
+        })
+    );
+  }
+
+  const value: GitHubFlowContext = {
+    attempt,
+    botConfig,
+    setBotConfig,
+    repoRules,
+    setRepoRules,
+    addEmptyRepoRule,
+    tokenName,
+    createBot,
+    resetAttempt,
+  };
+
+  return (
+    <gitHubFlowContext.Provider value={value}>
+      {children}
+    </gitHubFlowContext.Provider>
+  );
+}
+
+export function useGitHubFlow(): GitHubFlowContext {
+  return useContext(gitHubFlowContext);
+}
+
+export type RefTypeOption = Option<RefType | ''>;
+
+export type Rule = {
+  workflowName: string;
+  environment: string;
+  ref: string;
+  refType: RefTypeOption;
+  repoAddress: string;
+  actor: string;
+};
+
+export const defaultRule: Rule = {
+  workflowName: '',
+  environment: '',
+  ref: '',
+  refType: { label: 'any', value: '' },
+  repoAddress: '',
+  actor: '',
+};
+
+/**
+ * Parses the GitHub repository URL and returns the repository name and
+ * its owner's name. Throws errors if parsing the URL fails or
+ * the URL doesn't contains the expected format.
+ * @param repoAddr repository address (with or without protocl)
+ * @returns owner and repository name
+ */
+export function parseRepoAddress(repoAddr: string): {
+  host: string;
+  owner: string;
+  repository: string;
+} {
+  // add protocol if it is missing
+  if (!repoAddr.startsWith('http://') && !repoAddr.startsWith('https://')) {
+    repoAddr = `https://${repoAddr}`;
+  }
+
+  let url;
+  try {
+    url = new URL(repoAddr);
+  } catch (e) {
+    throw new Error('Must be a valid URL');
+  }
+
+  const paths = url.pathname.split('/');
+  // expected length is 3, since pathname starts with a /, so paths[0] should be empty
+  if (paths.length < 3) {
+    throw new Error(
+      'URL expected to be in the format https://<host>/<owner>/<repository>'
+    );
+  }
+
+  const owner = paths[1];
+  const repository = paths[2];
+  if (owner.trim() === '' || repository.trim() == '') {
+    throw new Error(
+      'URL expected to be in the format https://<host>/<owner>/<repository>'
+    );
+  }
+
+  return {
+    host: url.host,
+    owner,
+    repository,
+  };
+}
+
+function getRoleYaml(botName: string, labels: ResourceLabel[], login): string {
+  const labelsStanza = labels.map(
+    label => `'${label.name}': '${label.value}'\n`
+  );
+  const timestamp = new Date().getTime();
+
+  return `kind: role
+metadata:
+  name: bot-${botName}-role-${timestamp}
```

Comment:
Any suggestions for the role name?

Reply:
We could ask folks to name the integration and then use this in the name of the bot, role and token?